README.txt
License
THIS SCRIPT IS PROVIDED TO YOU "AS IS." TO THE EXTENT PERMITTED BY LAW, QUALYS HEREBY DISCLAIMS ALL WARRANTIES AND LIABILITY FOR THE PROVISION OR USE OF THIS SCRIPT. 
IN NO EVENT SHALL THESE SCRIPTS BE DEEMED TO BE CLOUD SERVICES AS PROVIDED BY QUALYS

Log4j Vulnerability Scanner Shell Script:

Description:
This shell script intends to collect necessary details and help detect CVE-2021-44228 and CVE-2021-45046 vulnerabilities reported in Log4j.
The script will scan the entire filesystem, including archives for the Java class that indicates the Java application contains a vulnerable Log4j library.
Once Log4j QID is introduced in Qualys VM signatures, the output file generated by this script will serve as a data point to assess and report the QID during agent VM scan.

Usage: 
Supported platforms: Linux(RHEL, CentOS, Ubuntu, Debian, Amazon Linux, and OEL), MacOS, AIX, and Solaris
Supported architectures: x64, ARM(Linux)

How to run the script? 
1) Create and save a script (.sh) file using any text editor.
2) Execute the script file using shell script. Use the following commands:

LINUX: (detects WARs, EARs, ZIPs, and nested JARs, located under linux folder) 
sh <script-name>.sh [base_dir] [network_filesystem_scan<true/false>]
Here, <script-name> is actual script name.
(default: [base_dir]=/ [network_filesystem_scan]=false)
For example: sh ./log4j_findings.sh /home false

MacOS/AIX/Solaris: (Beta version, detects JARs only, located under unix folder) 
sh <script-name>.sh [base_dir]
Here, <script-name> is actual script name.
(default: [base_dir]=/ )
For example: sh ./log4j_findings_unix.sh /home

The script’s standard output will be redirected to:
Linux: /usr/local/qualys/cloud-agent/log4j_findings.stdout
MacOS: /Library/Application\ Support/QualysCloudAgent/Data/log4j_findings.stdout
AIX/Solaris: /opt/qualys/cloud-agent/log4j_findings.stdout

Any error occurring during its execution is redirected to:
Linux: /usr/local/qualys/cloud-agent/log4j_findings.stderr
MacOS: /Library/Application\ Support/QualysCloudAgent/Data/log4j_findings.stderr
AIX/Solaris: /opt/qualys/cloud-agent/log4j_findings.stderr 
(This file will also contain command execution start time and status on completing the run)

If the agent is not installed, the script will create the following directory and dump the standard output and error in files within it. 
However, no status information will be written in the error file if its execution is aborted.
LINUX: /usr/local/qualys/cloud-agent/ 
MACOS: /Library/Application\ Support/QualysCloudAgent/Data/  
AIX/Solaris: /opt/qualys/cloud-agent/ 

The following details are shown in the output: 
Source: Path of pom.xml within a jar file
JNDI class found or not found status
Path of the jar
log4j version

Sample output:(/usr/local/qualys/cloud-agent/log4j_findings.stdout) 
Source: META-INF/maven/org.slf4j/slf4j-log4j12/pom.xml META-INF/maven/log4j/log4j/pom.xml
JNDI-Class: JNDI Class Not Found
Path= /root/TestFiles_log4j/TestFiles/kafka-producer-intellij.jar
log4j Unknown
------------------------------------------------------------------------
Source: META-INF/maven/org.apache.logging.log4j/log4j-api/pom.xml META-INF/maven/org.apache.logging.log4j/log4j-core/pom.xml
JNDI-Class: JNDI Class Found
Path= /root/TestFiles_log4j/TestFiles/ProjWithVulLog4j-1.0-SNAPSHOT-jar-with-dependencies.jar
log4j 2.13.0
------------------------------------------------------------------------
Source: META-INF/maven/log4j/log4j/pom.xml
JNDI-Class: JNDI Class Not Found
Path= /usr/share/java/log4j.jar
log4j Unknown
------------------------------------------------------------------------

Sample error file: (/usr/local/qualys/cloud-agent/log4j_findings.stderr)
scanning started for log4j jar
Thu Dec 16 17:47:36 IST 2021
Run status : Success


CHANGES.txt
Linux:

Version: 2.4
- Added changes to terminate the manual execution of the script if agent's inventory scan is in progress to prohibit overriding of the results into output file if happened simultaneously.

Version: 2.3
- Enhanced script to not to overwrite scan results when scan is disabled.

Version: 2.2
- Added support for scanning nested JARs
- Fixed issue related filesnames containing spaces in between in Ubuntu.

Version: 2.1
- Fixed the issue of not scanning all mounted drives by removing  "-xdev" option from the "find" command.
- Fixed the issue of failing to traverse the directory path with spaces embedded.

Version: 2.0
- Added support for scanning of JAR, WAR, EAR and ZIP files.

Version: 1.0
- Added support for scanning of JAR files.
---------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------

Unix script:

Version: 1.0
- Added support for scanning of JAR files.


unix/log4j_findings_unix.sh
#!/bin/sh

if [ $# -eq 0 ]; then
	BASEDIR="/";
elif [ $# -eq 1 ]; then
	BASEDIR=$1;
	if [ ! -d $BASEDIR ]; then
		echo "Please enter valid directory path";
		exit 1;
	fi;
else
	echo "Too many parameters passed in."
	echo "sh ./log4j_findings_unix.sh [base_dir] "
	echo "example: sh ./log4j_findings_unix.sh /home "
	echo "(default: [base_dir]=/ )" 
	exit 1
fi

log4j()
{
	echo "Script version: 1.0 (scans JAR files only)";	
	
	if [ "${OS}" = "AIX" ] || [ "${OS}" = "SunOS" ]; then
		echo "Scanning started.." > /opt/qualys/cloud-agent/log4j_findings.stderr;
		date >> /opt/qualys/cloud-agent/log4j_findings.stderr;
	elif [ "${OS}" = "Darwin" ]; then
		echo "Scanning started.." > /Library/Application\ Support/QualysCloudAgent/Data/log4j_findings.stderr ;
		date >> /Library/Application\ Support/QualysCloudAgent/Data/log4j_findings.stderr;
	fi;

	id=`id`
	if ! (echo $id | grep "uid=0") > /dev/null; then
		echo "Please run the script as root user for complete results";
	fi;
	
	log4j_exists=0;
	if [ ! -d /tmp/log4j_jar ]; then
		mkdir -p /tmp/log4j_jar;
	fi;	
	
	cd /tmp/log4j_jar 2>/dev/null	
        oldIFS=$IFS;	
	jars=$(find ${BASEDIR} -name "*.jar" -type f 2>/dev/null)
       IFS='
'	
	for i in $jars;	do
		if test=$(jar -tf $i | grep "[l]og4j-core" | grep "pom.xml" 2>/dev/null); then
			log4j_exists=1;	
			if jar -tf $i | grep "JndiLookup.class" >/dev/null; then
				jdi="JNDI Class Found"
			else
				jdi="JNDI Class Not Found"
			fi
			echo "Source: "$test
			echo "JNDI-Class: "$jdi
			echo 'Path= '$i;
			jar -xf $i > /dev/null
			if [ "${OS}" = "SunOS" ]; then
				ve=`cat $test | head -30 | ggrep -A 1 '<artifactId>log4j</artifactId>'| cut -d ">" -f 2 | cut -d "<" -f 1 | tail -1`
			else
				ve=`cat $test 2>/dev/null | grep -E "<artifactId>log4j</artifactId>
.+?<version>"| cut  -d ">" -f 2 | cut -d "<" -f 1 | head -2|tail -1`;
			fi;
			if [ -z "$ve" ]; then 
				echo 'log4j Unknown'; 
			else 
				echo 'log4j '$ve;
			fi;
			echo "------------------------------------------------------------------------"
			rm -rf /tmp/log4j_jar/*
		fi
	done
	IFS=$oldIFS;
	if [ $log4j_exists -eq 0 ]; then
		echo "No log4j jars found on the system , exiting now.";
	fi;
	
	if [ "${OS}" = "AIX" ] || [ "${OS}" = "SunOS" ]; then
		echo "Run Status : Success" >> /opt/qualys/cloud-agent/log4j_findings.stderr;
	elif [ "${OS}" = "Darwin" ]; then
		echo "Run Status : Success" >> /Library/Application\ Support/QualysCloudAgent/Data/log4j_findings.stderr;
	fi;
}

OS=`uname -s`

if [ "${OS}" = "AIX" ] || [ "${OS}" = "SunOS" ]; then
	if [ ! -d "/opt/qualys/cloud-agent/" ]; then
		mkdir -p "/opt/qualys/cloud-agent/";
		chmod 750 "/opt/qualys/cloud-agent/";
	fi; 
	if [ ! -f "/opt/qualys/cloud-agent/log4j_findings_disabled" ]; then
		log4j > /opt/qualys/cloud-agent/log4j_findings.stdout 2>/opt/qualys/cloud-agent/log4j_findings.stderr;
	else
		rm -rf /opt/qualys/cloud-agent/log4j_findings.stdout; 
		echo "Flag is disabled, skipping command execution" > /opt/qualys/cloud-agent/log4j_findings.stderr;
	fi;
elif [ "${OS}" = "Darwin" ]; then
	if [ ! -d /Library/Application\ Support/QualysCloudAgent/Data/ ]; then
		mkdir -p /Library/Application\ Support/QualysCloudAgent/Data/;
		chmod 750 /Library/Application\ Support/QualysCloudAgent/Data/;
	fi;
	if [ ! -f /Library/Application\ Support/QualysCloudAgent/Data/log4j_findings_disabled ]; then
		log4j > /Library/Application\ Support/QualysCloudAgent/Data/log4j_findings.stdout 2>/Library/Application\ Support/QualysCloudAgent/Data/log4j_findings.stderr;
	else
		rm -rf /Library/Application\ Support/QualysCloudAgent/Data/log4j_findings.stdout;
		echo "Flag is disabled, skipping command execution" > /Library/Application\ Support/QualysCloudAgent/Data/log4j_findings.stderr;
	fi;
else
	echo "Unsupported platform: ${OS}, script supports only AIX, MacOS and Solaris platforms.";
fi;



linux/log4j_findings.sh
#!/bin/sh

if [ $# -eq 0 ]; then
	BASEDIR="/"
	NETDIR_SCAN=false
elif [ $# -eq 1 ]; then
	BASEDIR=$1
	if [ ! -d $BASEDIR ];then
		echo "Please enter valid directory path";
		exit 1;
	fi;
	NETDIR_SCAN=false
elif [ $# -eq 2 ]; then
	BASEDIR=$1
	NETDIR_SCAN=$2
	if [ ! -d $BASEDIR ]; then
		echo "Please enter valid directory path";
		exit 1;
	fi;
else
	echo "Too many parameters passed in."
	echo "sh ./log4j_findings.sh [base_dir] [network_filesystem_scan<true/false>]"
	echo "example: sh ./log4j_findings.sh /home false"
	echo "(default: [base_dir]=/ [network_filesystem_scan]=false)"
	exit 1
fi

handle_war_ear_zip()
{
	war_file=$1
	if jar1=`zip -sf  $war_file | awk '{print $NF}'| grep -i ".jar" 2> /dev/null `;then
		rm -rf /tmp/log4j_for_extract/
		mkdir /tmp/log4j_for_extract;
		unzip -o -d /tmp/log4j_for_extract/ $war_file > /dev/null
	fi;
	jars=`find /tmp/log4j_for_extract -type f -regextype posix-egrep -iregex ".+\.(jar)$"  2> /dev/null`; 
	for i in $jars; do 		
		IFS='
		'
		handle_jar $i $war_file	
	done;
	rm -rf /tmp/log4j_for_extract/
}

handle_jar_without_zip()
{	
	jar_file=$1;	
	echo "Zip/Unzip utility not present on the system, showing limited results for " $jar_file " only in output file" >> /usr/local/qualys/cloud-agent/log4j_findings.stderr;

	var=`echo $jar_file | grep -i "log4j.*jar"` 2> /dev/null; 
	if [ ! -z "$var" ]; then 
		log4j_exists=1;				
		echo 'Path= '$jar_file; 
		## Fetch log4j version from jar
		ver=`echo $jar_file | grep -o '[^\/]*$' | grep -oE "([0-9]+\.[0-9]+\.[0-9]+-[a-zA-Z0-9]*[0-9]*|[0-9]+\.[0-9]+-[a-zA-Z0-9]+[0-9]*|[0-9]+\.[0-9]+\.[0-9]+|[0-9]+\.[0-9]+)" | tail -1` 2> /dev/null; 
		if [ -z "$ver" ]; then 
			echo 'log4j Unknown'; 
		else 
			echo 'log4j '$ver; 
		fi; 
		echo "------------------------------------------------------------------------"; 
	else 
		injars=`(jar -tf $jar_file | grep -i "log4j.*jar") 2> /dev/null`; 
		for j in $injars ; do 
			if [ ! -z "$j" ]; then 						
				log4j_exists=1;
				echo 'Path= '$j; 
				## Fetch log4j version from jar
				ver1=`echo $j | grep -o '[^\/]*$' | grep -oE "([0-9]+\.[0-9]+\.[0-9]+-[a-zA-Z0-9]*[0-9]*|[0-9]+\.[0-9]+-[a-zA-Z0-9]+[0-9]*|[0-9]+\.[0-9]+\.[0-9]+|[0-9]+\.[0-9]+)" | tail -1` 2> /dev/null; 
				if [ -z "$ver1" ]; then 
					echo 'log4j Unknown'; 
				else 
					echo 'log4j '$ver1; 
				fi; 
			fi; 
			echo "------------------------------------------------------------------------"; 
		done;
	fi;
}

handle_jar_with_zip()
{
	jar_file=$1
	war_file=$2	
	if zip -sf $jar_file | grep "JndiLookup.class" >/dev/null; then 
		jdi="JNDI Class Found";
	else 
		jdi="JNDI Class Not Found";
	fi;	
	## Checking JNDI-Class value from jar file
	if test=`zip -sf $jar_file | grep -i "log4j-core" | grep "pom.xml"`;then 
		IFS=$oldIFS;
		echo "Source: "$test;
		log4j_exists=1;
		## Reading file pom.xml to fetch log4j version
		echo "JNDI-Class: "$jdi;		
		IFS='
		'		
		if [ ! -z "$war_file" ];then
			p=`echo $jar_file | sed -n 's|^/tmp/log4j_for_extract/||p' `;
			echo 'Path= '$war_file'/'$p
		else 
			echo 'Path= '$jar_file
		fi
		IFS=$oldIFS;
		ve=`unzip -p $jar_file $test 2> /dev/null | grep -Pzo "<artifactId>log4j</artifactId>\s*<version>.+?</version>"| cut -d ">" -f 2 | cut -d "<" -f 1 | head -2|awk 'ORS=NR%3?FS:RS'`;
		if [ -z "$ve" ]; then 
			echo 'log4j Unknown'; 
		else 
			echo $ve; 
		fi;
		echo "------------------------------------------------------------------------";
	fi;
}

handle_jar()
{
  if [ "$isZip" -eq 0 ] && [ "$isUnZip" -eq 0 ]; then         
	handle_jar_with_zip "$@";  
  else
	handle_jar_without_zip "$@";  
  fi;
}

log4j()
{
    echo "Script version: 2.3 (scans jar/war/ear/zip files)" ;
    echo "Scanning started.." > /usr/local/qualys/cloud-agent/log4j_findings.stderr ;
    date >> /usr/local/qualys/cloud-agent/log4j_findings.stderr ;    
    id=`id`;
    if ! (echo $id | grep "uid=0")>/dev/null; then 
        echo "Please run the script as root user for complete results";
    fi;    
    zip -v 2> /dev/null 1> /dev/null;
    isZip=$?;
    unzip -v 2> /dev/null 1> /dev/null;
    isUnZip=$?;
    log4j_exists=0;
    oldIFS=$IFS;

    # Change to a network filesystem only scan if 2nd parameter is true. network filesystem scan command
    # does not use '!' flags
    if [ $NETDIR_SCAN = true ];then
        jars=`find ${BASEDIR} -type f -regextype posix-egrep -iregex ".+\.(jar|war|ear|zip)$" 2> /dev/null`; 
    else
	jars=`find ${BASEDIR} -type f -regextype posix-egrep -iregex ".+\.(jar|war|ear|zip)$"  ! -fstype nfs ! -fstype nfs4 ! -fstype cifs ! -fstype smbfs ! -fstype gfs ! -fstype gfs2 ! -fstype safenetfs ! -fstype secfs ! -fstype gpfs ! -fstype smb2 ! -fstype vxfs ! -fstype vxodmfs ! -fstype afs -print 2>/dev/null`;
    fi 
	
    	IFS='
	'
	for i in $jars ; do
		if `echo $i | grep -q ".jar"`; then		
			if nested=`zip -sf $i | awk '{print $NF}'| grep -i ".jar" 2>/dev/null`;then
				handle_war_ear_zip $i
			else
				handle_jar $i
			fi;		
		else
			if [ "$isZip" -eq 0 ] && [ "$isUnZip" -eq 0 ];then 
				handle_war_ear_zip $i
			else
				echo "Zip/Unzip utility not present on the system, skipping processing of file: "$i >> /usr/local/qualys/cloud-agent/log4j_findings.stderr;
			fi
		fi
    		IFS='
		'
	done
	IFS=$oldIFS;
	if [ $log4j_exists -eq 0 ]; then
		echo "No log4j jars found on the system , exiting now.";
	fi;
    echo "Run status : Success" >> /usr/local/qualys/cloud-agent/log4j_findings.stderr;
};

isInventoryScanInProgress=`ps -ef | grep /usr/local/qualys/cloud-agent/ChangeListInventory.db | grep -v grep`
if [ ! -z "$isInventoryScanInProgress" ]; then 
    echo "Inventory scan is in progress, exiting the manual execution of the script for now..."
    exit 0
fi

if [ ! -d "/usr/local/qualys/cloud-agent/" ]; then 
    mkdir -p "/usr/local/qualys/cloud-agent/";
    chmod 750 "/usr/local/qualys/cloud-agent/";
fi; 

if [ ! -f "/usr/local/qualys/cloud-agent/log4j_findings_disabled" ]; then 
    log4j > /usr/local/qualys/cloud-agent/log4j_findings.stdout 2>/usr/local/qualys/cloud-agent/log4j_findings.stderr;
else 
    prevFlag=`grep "Flag is disabled, skipping command execution" /usr/local/qualys/cloud-agent/log4j_findings.stderr 2>/dev/null`
    if [ -z "$prevFlag" ]; then    
       echo "Checking log4j flag.." >> /usr/local/qualys/cloud-agent/log4j_findings.stderr ;       
       date >> /usr/local/qualys/cloud-agent/log4j_findings.stderr ;
       echo "Flag is disabled, skipping command execution" >> /usr/local/qualys/cloud-agent/log4j_findings.stderr;
    fi;
fi;

